File: android/app/src/main/kotlin/dev/agentdeck/terrarium/creature/BubbleSystem.kt
package dev.agentdeck.terrarium.creature

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.drawscope.DrawScope
import dev.agentdeck.terrarium.EnvironmentVisualState
import dev.agentdeck.terrarium.TerrariumColors
import dev.agentdeck.terrarium.TerrariumTiming
import kotlin.math.PI
import kotlin.math.sin
import kotlin.random.Random

/**
 * Particle engine for bubbles rising from the bottom.
 * Ring buffer (max 50). Spawn rate depends on environment state.
 */
class BubbleSystem : Creature {

    private data class Bubble(
        var x: Float,
        var y: Float,
        var radius: Float,
        var speed: Float,
        var wobblePhase: Float,
        var wobbleAmp: Float,
        var alpha: Float,
        var alive: Boolean = true,
    )

    private val bubbles = Array(MAX_BUBBLES) {
        Bubble(0f, 0f, 0f, 0f, 0f, 0f, 0f, alive = false)
    }
    private var nextSlot = 0
    private var timeSinceSpawn = 0f
    private var envState by mutableStateOf(EnvironmentVisualState.CALM)
    private var time by mutableFloatStateOf(0f)

    fun setState(state: EnvironmentVisualState) {
        envState = state
    }

    override fun update(dt: Float) {
        time += dt
        timeSinceSpawn += dt * 1000f

        val spawnInterval = when (envState) {
            EnvironmentVisualState.DARK -> Float.MAX_VALUE // no bubbles
            EnvironmentVisualState.CALM -> TerrariumTiming.CALM_SPAWN_INTERVAL_MS
            EnvironmentVisualState.ACTIVE -> TerrariumTiming.ACTIVE_SPAWN_INTERVAL_MS
            EnvironmentVisualState.ALERT -> TerrariumTiming.ACTIVE_SPAWN_INTERVAL_MS * 1.5f
        }

        // Spawn new bubbles
        while (timeSinceSpawn >= spawnInterval) {
            timeSinceSpawn -= spawnInterval
            spawnBubble()
        }

        // Update existing bubbles
        for (bubble in bubbles) {
            if (!bubble.alive) continue

            bubble.y -= bubble.speed * dt
            bubble.x += sin(time * TerrariumTiming.BUBBLE_WOBBLE_SPEED + bubble.wobblePhase) *
                bubble.wobbleAmp * dt

            // Fade out near top
            if (bubble.y < 0.1f) {
                bubble.alpha = (bubble.y / 0.1f).coerceIn(0f, 1f)
            }

            // Kill if off screen
            if (bubble.y < -0.02f) {
                bubble.alive = false
            }
        }
    }

    override fun draw(scope: DrawScope) {
        val w = scope.size.width
        val h = scope.size.height

        for (bubble in bubbles) {
            if (!bubble.alive) continue

            val screenX = bubble.x * w
            val screenY = bubble.y * h
            val screenRadius = bubble.radius * w

            // Bubble body
            scope.drawCircle(
                color = TerrariumColors.BubbleWhite.copy(alpha = bubble.alpha * 0.3f),
                radius = screenRadius,
                center = Offset(screenX, screenY),
            )

            // Bubble highlight (upper-left)
            scope.drawCircle(
                color = TerrariumColors.BubbleHighlight.copy(alpha = bubble.alpha * 0.5f),
                radius = screenRadius * 0.3f,
                center = Offset(
                    screenX - screenRadius * 0.25f,
                    screenY - screenRadius * 0.25f,
                ),
            )
        }
    }

    private fun spawnBubble() {
        val bubble = bubbles[nextSlot]
        nextSlot = (nextSlot + 1) % MAX_BUBBLES

        val isError = envState == EnvironmentVisualState.ALERT

        bubble.x = Random.nextFloat() * 0.8f + 0.1f // 10%-90% width
        bubble.y = 0.95f + Random.nextFloat() * 0.05f // bottom
        bubble.radius = if (isError) {
            Random.nextFloat() * 0.008f + 0.005f
        } else {
            Random.nextFloat() * 0.005f + 0.002f
        }
        bubble.speed = TerrariumTiming.BUBBLE_RISE_SPEED * (0.7f + Random.nextFloat() * 0.6f)
        bubble.wobblePhase = Random.nextFloat() * 2f * PI.toFloat()
        bubble.wobbleAmp = Random.nextFloat() * 0.02f + 0.005f
        bubble.alpha = 1f
        bubble.alive = true
    }

    companion object {
        private const val MAX_BUBBLES = 50
    }
}

File: apple/AgentDeck/Terrarium/Creatures/BubbleSystem.swift
// BubbleSystem.swift — Rising bubbles + pop burst effects
// Ported from android BubbleSystem.kt

import SwiftUI

final class BubbleSystem {
    // MARK: - Bubble

    private struct Bubble {
        var x: Float = 0
        var y: Float = 0
        var radius: Float = 0
        var speed: Float = 0
        var wobblePhase: Float = 0
        var wobbleAmp: Float = 0
        var alpha: Float = 0
        var alive: Bool = false
        // Pop burst fields
        var popping: Bool = false
        var popProgress: Float = 0
        var popOriginX: Float = 0
        var popOriginY: Float = 0
        var popAngle: Float = 0
        var popDistance: Float = 0
    }

    // MARK: - State

    private static let maxBubbles = 70
    private static let popSpeed: Float = 2.5

    private var bubbles: [Bubble]
    private var nextSlot = 0
    private var timeSinceSpawn: Float = 0
    private var time: Float = 0
    private var envState: EnvironmentVisualState = .calm

    init() {
        bubbles = Array(repeating: Bubble(), count: Self.maxBubbles)
    }

    func setState(_ state: EnvironmentVisualState) {
        envState = state
    }

    // MARK: - Pop Burst

    /// Emit a radial burst of small bubbles — triggered when leaving ASKING state.
    func emitPopBurst(x: Float, y: Float, count: Int = 10) {
        let angleStep = Float.pi * 2 / Float(count)
        for i in 0..<count {
            let angle = angleStep * Float(i) + Float.random(in: 0..<angleStep * 0.5)
            bubbles[nextSlot] = Bubble(
                x: x, y: y,
                radius: Float.random(in: 0.002...0.006),
                speed: 0,
                wobblePhase: Float.random(in: 0...Float.pi * 2),
                wobbleAmp: Float.random(in: 0.003...0.013),
                alpha: 0.9,
                alive: true,
                popping: true,
                popProgress: 0,
                popOriginX: x,
                popOriginY: y,
                popAngle: angle,
                popDistance: 0.03 + Float.random(in: 0...0.04)
            )
            nextSlot = (nextSlot + 1) % Self.maxBubbles
        }
    }

    /// Emit small bubbles from a creature's position (50% smaller, 70% speed).
    func emitCreatureBubbles(x: Float, y: Float, count: Int) {
        for _ in 0..<count {
            bubbles[nextSlot] = Bubble(
                x: x + Float.random(in: -0.01...0.01),
                y: y - 0.01,
                radius: Float.random(in: 0.001...0.004),
                speed: TerrariumTiming.bubbleRiseSpeed * (0.7 + Float.random(in: 0...0.6)) * 0.7,
                wobblePhase: Float.random(in: 0...Float.pi * 2),
                wobbleAmp: Float.random(in: 0.003...0.018),
                alpha: 0.8,
                alive: true
            )
            nextSlot = (nextSlot + 1) % Self.maxBubbles
        }
    }

    // MARK: - Update

    func update(dt: Float) {
        time += dt
        timeSinceSpawn += dt * 1000

        let spawnInterval: Float = switch envState {
        case .dark: Float.greatestFiniteMagnitude
        case .calm: 2000  // CALM_SPAWN_INTERVAL_MS
        case .active: 300  // ACTIVE_SPAWN_INTERVAL_MS
        case .alert: 450   // ACTIVE * 1.5
        }

        // Spawn new bubbles from bottom
        while timeSinceSpawn >= spawnInterval {
            timeSinceSpawn -= spawnInterval
            spawnBubble()
        }

        // Update existing bubbles
        for i in bubbles.indices {
            guard bubbles[i].alive else { continue }

            if bubbles[i].popping {
                bubbles[i].popProgress += dt * Self.popSpeed
                if bubbles[i].popProgress >= 1.0 {
                    // Transition to normal rising bubble
                    bubbles[i].popping = false
                    bubbles[i].speed = TerrariumTiming.bubbleRiseSpeed * (0.5 + Float.random(in: 0...0.4))
                } else {
                    // Radial expansion with ease-out
                    let ease = 1 - pow(1 - bubbles[i].popProgress, 2)
                    bubbles[i].x = bubbles[i].popOriginX + cos(bubbles[i].popAngle) * bubbles[i].popDistance * ease
                    bubbles[i].y = bubbles[i].popOriginY + sin(bubbles[i].popAngle) * bubbles[i].popDistance * ease
                    bubbles[i].radius = max(0.001, bubbles[i].radius * (1 - bubbles[i].popProgress * 0.3))
                    bubbles[i].alpha = max(0.3, 1 - bubbles[i].popProgress * 0.4)
                }
                continue
            }

            bubbles[i].y -= bubbles[i].speed * dt
            bubbles[i].x += sin(time * TerrariumTiming.bubbleWobbleSpeed + bubbles[i].wobblePhase) *
                bubbles[i].wobbleAmp * dt

            // Fade out near top
            if bubbles[i].y < 0.1 {
                bubbles[i].alpha = min(1, max(0, bubbles[i].y / 0.1))
            }

            // Kill if off screen
            if bubbles[i].y < -0.02 {
                bubbles[i].alive = false
            }
        }
    }

    private func spawnBubble() {
        let isError = envState == .alert
        bubbles[nextSlot] = Bubble(
            x: Float.random(in: 0.1...0.9),
            y: 0.95 + Float.random(in: 0...0.05),
            radius: isError ? Float.random(in: 0.005...0.013) : Float.random(in: 0.002...0.007),
            speed: TerrariumTiming.bubbleRiseSpeed * (0.7 + Float.random(in: 0...0.6)),
            wobblePhase: Float.random(in: 0...Float.pi * 2),
            wobbleAmp: Float.random(in: 0.005...0.025),
            alpha: 1.0,
            alive: true
        )
        nextSlot = (nextSlot + 1) % Self.maxBubbles
    }

    // MARK: - Draw

    func draw(context: inout GraphicsContext, size: CGSize) {
        let w = Float(size.width)
        let h = Float(size.height)

        for bubble in bubbles where bubble.alive {
            let screenX = CGFloat(bubble.x * w)
            let screenY = CGFloat(bubble.y * h)
            let screenR = CGFloat(bubble.radius * w)

            // Bubble body
            let bodyRect = CGRect(x: screenX - screenR, y: screenY - screenR,
                                  width: screenR * 2, height: screenR * 2)
            context.fill(Path(ellipseIn: bodyRect),
                         with: .color(Color.white.opacity(Double(bubble.alpha) * 0.3)))

            // Highlight (upper-left)
            let hlR = screenR * 0.3
            let hlRect = CGRect(x: screenX - hlR - screenR * 0.25,
                                y: screenY - hlR - screenR * 0.25,
                                width: hlR * 2, height: hlR * 2)
            context.fill(Path(ellipseIn: hlRect),
                         with: .color(Color.white.opacity(Double(bubble.alpha) * 0.5)))
        }
    }
}
